snuba/utils/streams/kafka.py

```diff
@@ -29,7 +29,20 @@ class TopicPartition(NamedTuple):
     partition: int
 
 
-KafkaMessage = Message[TopicPartition, int, bytes]
+class KafkaMessage(Message[TopicPartition, int, bytes]):
+    def __init__(
+        self,
+        stream: TopicPartition,
+        offset: int,
+        value: bytes,
+        next_offset: Optional[int] = None,
+    ) -> None:
+        super().__init__(
+            stream,
+            offset,
+            value,
+            next_offset if next_offset is not None else offset + 1,
```

Comment:
Can you envision a use case where the user would pass a next_offset as input parameter instead of relying on this class to calculate it ? It seems a bit of a leaky abstraction since it allows the callsite to know about offset calculation

Reply:
This is a good point — I was thinking that this needed to be a constructor parameter since it is part of the `Message` dataclass. Thinking about it more, this shouldn't actually be an attribute on the dataclass at all — the `Message` should implement an abstract method `get_next_offset() -> TOffset` that provides the next offset.

If the next offset can be derived mathematically for a specific implementation from the current offset (like it is here) there's no reason to make it part of the constructor. If a specific implementation requires that the next offset needs to be defined by some other means, it's free to implement it as a constructor parameter in it's implementation of `poll`.